### DISCUSSION TITLE
Remote Debugging with PPSSPP

### DISCUSSION THREAD

**Discussion by Nemoumbra** [kotcrab/ghidra-allegrex#31] [General]
I'd really love to help with moving the PPSSPP debugger integration out of the "beta" stage. It can turn out to be very helpful for a lot of modders/researchers.
Unfortunately, I don't know Java or Kotlin, but I can help with testing! I've personally analyzed more than 30 PSP games in Ghidra and I consider myself a somewhat experienced PPSSPP and Ghidra user. I suggest that we use this discussion thread for talking about anything related to the debugging workflow (maybe suggestions/bugs, etc).
I don't mind even writing a tutorial at some point.

**Comment by kotcrab** [comment]
Personally I haven't used it pretty much at all after making it. It doesn't provide much over the built-in debugger in PPSSPP and the slowness and clunkiness of the Ghidra UI really doesn't help.

**Reply by Nemoumbra** [reply]
Oh, I disagree! The Decompiler is very much "over" the PPSSPP functionality!

**Reply by kotcrab** [reply]
I was talking about the debugger. The decompiler has some basic integration with it but it's pretty lacking compared to what could it be. A lot of untapped potential here.

**Comment by Nemoumbra** [comment]
Process > Memory > Uncached Kernel Memory
Selecting this causes PPSSPP to die.
https://github.com/hrydgard/ppsspp/issues/17851

**Reply by Nemoumbra** [reply]
Fixed by https://github.com/hrydgard/ppsspp/pull/18104

**Comment by Nemoumbra** [comment]
Alright, question number one...
That's what PPSSPP shows: 
![image](https://github.com/kotcrab/ghidra-allegrex/assets/62447396/654a0e06-f2f2-496d-9dac-c4b64a7c51ff)

The current thread is "root", pc = 08869238.
Now that's what Ghidra shows:
![image](https://github.com/kotcrab/ghidra-allegrex/assets/62447396/a8020164-2205-4c8b-99e8-766e65fc5d82)

The "Registers" view doesn't show me the registers for the thread "root". Why? Can I switch to the other thread there? I can surely find the values in the "Objects" view (on the left of the image), but I don't want to.

**Reply by kotcrab** [reply]
You can double click the thread in that list.

**Reply by Nemoumbra** [reply]
Thanks, that worked! I was confused before because clicking on the name prompts you to enter a different name. There are only a few places where you can click to achieve this.

**Comment by Nemoumbra** [comment]
Second question...
How can I goto current PC in Ghidra? PPSSPP has paused the execution due to a memory breakpoint firing. The Debugger window in PPSSPP immediately jumped to PC, but not Ghidra. It seems to be a bit dumb with detecting what thread is currently executed (sgx-psp-pcm-th in reality and idle0 according to Ghidra). Maybe something can be made about it? Please try this for yourself, maybe there's a random hidden button I'm missing.

**Reply by kotcrab** [reply]
I noticed it too, maybe Ghidra needs some hint provided about current thread when breakpoints triggers.

**Reply by Nemoumbra** [reply]
Oh yeah. We can either make a discussion or search the standard GDB debugger's code.

**Comment by Nemoumbra** [comment]
I have found another peculiarity. Let's enable Break on load, then connect the debugger and start the game. It looks like Ghidra thinks the game is running. I mean, this is how the UI looks like. I don't know how to start the game.

![image](https://github.com/kotcrab/ghidra-allegrex/assets/62447396/aa731028-96cf-49d4-a228-96292dffa0f9)
